### PR TITLE
Move gyro calibration from commander to events

### PR DESCRIPTION
**Pull request by sugnanprabhu**
This patch moves gyro calibration out of commander to events framework.
Some of the common code shared with other sensor calibration code is moved
to systemlib/calibration

This is one of the part of task mentioned in #8014

Currently the changes are tested on pixhawk using QGC and from commandline and compared the gyro calibration values with and without the changes.

### CODE REVIEW

**Comment by dagar** [comment]
It's hard to review through github, but so far it looks good!

CI failure is a minor clang-tidy issue.
image

File: src/modules/commander/commander.cpp [hunk 1]
```diff
@@ -449,7 +449,25 @@ int commander_main(int argc, char *argv[])
 			} else if (!strcmp(argv[2], "accel")) {
 				calib_ret = do_accel_calibration(&mavlink_log_pub);
 			} else if (!strcmp(argv[2], "gyro")) {
-				calib_ret = do_gyro_calibration(&mavlink_log_pub);
+				/* send vehicle_command to trigger gyro calibration */
+				struct vehicle_command_s cmd = {
```
Comment (dagar) [inline src/modules/commander/commander.cpp]: This is good for moving them incrementally, but once commander has all calibration code removed I think we can get rid it.

**Comment by dagar** [comment]
I think the notification stuff should stay private to commander (and needs a cleanup later).

LED control can be done via uORB like this -  https://github.com/PX4/Firmware/blob/master/src/modules/events/temperature_calibration/task.cpp#L343

tones can be written to the tone alarm device like this - https://github.com/PX4/Firmware/blob/master/src/modules/mavlink/mavlink_receiver.cpp#L1510

**Comment by LorenzMeier** [comment]
@sugnanprabhu Could you rebase?

**Comment by sugnanprabhu** [comment]
@LorenzMeier I have rebased the changes.

File: src/modules/events/calibration.cpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+#include <uORB/topics/calibration_status.h>
+
+#include "calibration.h"
+
+class Calibration;
+
+namespace calibration
+{
+Calibration 	*instance = nullptr;
+orb_advert_t _mavlink_log_pub = nullptr;
```
Comment (dagar) [inline src/modules/events/calibration.cpp]: Why keep these out of the class?

Comment (sugnanprabhu) [reply src/modules/events/calibration.cpp]: Idea is to have only one sensor calibration running at a time and avoid the re initialization of  these variables during multiple sensor calibration instance. We can choose to make these variable as a static members of Calibration class.

Similar approach is being followed in temperature calibration.
https://github.com/PX4/Firmware/blob/master/src/modules/events/temperature_calibration/task.cpp#L62

Comment (dagar) [reply src/modules/events/calibration.cpp]: Advertising these again is cheap and they shouldn't be left around after a calibration is finished even if you're about to do another one. They can be private in Calibration and unadvertised() in the destructor. 

If you want you can use this uORB Publication wrapper (https://github.com/PX4/Firmware/blob/master/src/modules/uORB/Publication.hpp#L123) to handle everything for you.

**Comment by dagar** [comment]
You can keep the calibration routines in with the calibration code for now, I don't think it belongs in systemlib (which should already be split in multiple libraries).

**Comment by sugnanprabhu** [comment]
@dagar moved the calibration routines back to commander. As discussed, some of the calibration code is still shared between the other sensor calibration in commander and gyro calibration in events, we can keep this way for now and move calibration routines to events with the movement of all calibration code.

**Comment by dagar** [comment]
Great! @PX4TestFlights could you give this a try on a few different boards? You don't need to do a full flight test, just set logging from boot and try the gyro calibration. With the logging running from boot we'll be able to compare the params before and after. Thanks!

**Comment by PX4TestFlights** [comment]
We'll test this out in the morning... what would be the potential issues
and expected behavior?

**Comment by santiago3dr** [comment]
tested on pixracer (V4)
gyro cal failed to start
https://logs.px4.io/plot_app?log=7ee149d6-b61f-43df-add3-ada78acfdad5

**Comment by sugnanprabhu** [comment]
@dagar I do not have pixracer to test here.
px4_task_spawn_cmd is failing, any idea what could be the issue?
https://github.com/PX4/Firmware/pull/8015/files#diff-d20167b763bb6f6b6ac105473dfa4931R97

@santiago3dr did you get a chance to test it on other boards?

**Comment by dagar** [comment]
I'll give it a try on my pixracer.

**Comment by santiago3dr** [comment]
calibration went through just fine on a pixhawk 3 pro (v4pro)
https://logs.px4.io/plot_app?log=b7dc9c78-b559-4947-b015-10b87fd4462b

**Comment by sugnanprabhu** [comment]
@santiago3dr I reduced the stacksize from 5800 to 3000 same as used by commander_low_prio_loop to make sure when other calibration routines are moved in next patches it still works.
Can you please give it a try on pixracer?

**Comment by r0gelion** [comment]
@sugnanprabhu 

tested on pixracer, calibration didn't work:
https://logs.px4.io/plot_app?log=184c0a00-46d4-4a7d-a7be-e4b9784c4b4f

**Comment by dagar** [comment]
We should get the error messages fixed here. https://github.com/PX4/Firmware/issues/8100

I have some ideas for things we can do with logger, etc to ensure there's always enough free memory to perform the calibration.

**Comment by r0gelion** [comment]
@sugnanprabhu @dagar 

The previous calibration was with USB connection, I calibrated again disconnecting USB, with telemetry connection and it worked.
https://logs.px4.io/plot_app?log=6c9378f6-f889-41fb-b71b-5c0155756599

**Comment by dagar** [comment]
Thanks @r0gelion 
This should be acceptable if we restore proper error messages. (again https://github.com/PX4/Firmware/issues/8100)
I'll also add a simple message to suggest disconnecting usb if you hit out of memory during calibration and it's connected.

**Comment by sugnanprabhu** [comment]
As moving the calibration incrementally is causing memory issue on pixracer, created a new PR moving all the calibration to events #8200

**Pull request closed without merging**